component/module/string/string.go

```diff
@@ -0,0 +1,193 @@
+}
+
+// Exports holds values which are exported from the run module.
+type Exports struct {
+	// Values exported from the running module.
+	Values map[string]any `river:"values,attr"`
+}
+
+// Component implements the module.string component.
+type Component struct {
+	opts component.Options
+	log  log.Logger
+	ctrl *flow.Flow
+
+	moduleArgs        atomic.Pointer[map[string]any]
+	file              atomic.Pointer[flow.File]
+	updateContollerCh chan struct{}
+
+	healthMut sync.RWMutex
+	health    component.Health
+}
+
+var (
+	_ component.Component       = (*Component)(nil)
+	_ component.HealthComponent = (*Component)(nil)
+	_ component.HTTPComponent   = (*Component)(nil)
+)
+
+// New creates a new module.string component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	// TODO(rfratto): replace these with a logger/tracer/registry which properly
+	// propagates data back to the parent.
+	flowLogger, _ := logging.New(os.Stderr, logging.Options{
+		Level:  logging.LevelDebug,
+		Format: logging.FormatLogfmt,
+	})
+	flowTracer, _ := tracing.New(tracing.DefaultOptions)
+	flowRegistry := prometheus.NewRegistry()
+
+	c := &Component{
+		opts: o,
+		log:  o.Logger,
+
+		ctrl: flow.New(flow.Options{
+			ControllerID: o.ID,
+			Logger:       flowLogger,
+			Tracer:       flowTracer,
+			Reg:          flowRegistry,
+
+			DataPath:       o.DataPath,
+			HTTPPathPrefix: o.HTTPPath,
+			HTTPListenAddr: o.HTTPListenAddr,
+
+			OnExportsChange: func(exports map[string]any) {
+				o.OnStateChange(Exports{Values: exports})
+			},
+		}),
+
+		updateContollerCh: make(chan struct{}, 1),
+	}
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+Loop:
+	for {
+		select {
+		case <-ctx.Done():
+			break Loop
+		case <-c.updateContollerCh:
+			err := c.ctrl.LoadFile(c.file.Load(), *c.moduleArgs.Load())
+			c.updateHealth(err)
+		}
+	}
+
+	return c.ctrl.Close()
+}
+
+func (c *Component) updateHealth(err error) {
+	c.healthMut.Lock()
+	defer c.healthMut.Unlock()
+
+	if err == nil {
+		c.health = component.Health{
+			Health:     component.HealthTypeHealthy,
+			Message:    "module updated",
+			UpdateTime: time.Now(),
+		}
+	} else {
+		c.health = component.Health{
+			Health:     component.HealthTypeUnhealthy,
+			Message:    err.Error(),
+			UpdateTime: time.Now(),
+		}
+	}
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+
+	f, err := flow.ReadFile(c.opts.ID, []byte(newArgs.Source))
+	if err != nil {
+		return err
+	}
+
+	// TODO(rfratto): sync exports with current set rather than updating it in
+	// full every time.
+	var emptyExports = map[string]any{}
+
+	for _, b := range f.ConfigBlocks {
+		if b.Name[0] == "export" {
+			emptyExports[b.Label] = nil
+		}
+	}
+	c.opts.OnStateChange(Exports{Values: emptyExports})
```

Comment:
I might be missing something, but why do we propagate an emptyExports here? Doesn't this mean we'll never inform the controller that it should propagate a change? Or is that something for upcoming PRs?

Reply:
(Note that the module loader component is taken out of the PR but the original commit is left in for referencing example usage of the controller)